### PR TITLE
chore: emulate `frameStartedNavigating` event

### PR DESCRIPTION
**Pull request by sadym-chromium**
Implementation of [Alternative 3. Rely on “Network.requestWillBeSent” CDP event](http://goto.google.com/webdriver:detect-navigation-started#bookmark=id.64balpqrmadv). Required for https://github.com/GoogleChromeLabs/chromium-bidi/issues/2856.

Emit `frameStartedNavigating` event on CdpTarget before `Network.requestWillBeSent`.

### CODE REVIEW

File: src/bidiMapper/modules/cdp/CdpTarget.ts
Comment (OrKoN) [inline src/bidiMapper/modules/cdp/CdpTarget.ts:62]: do we need to rename it? I think we are still toggling the CDP's network domain and not the bidi network domain.

File: src/bidiMapper/modules/cdp/CdpTarget.ts
Comment (OrKoN) [inline src/bidiMapper/modules/cdp/CdpTarget.ts:447]: could we force enable here instead of #unblock above?

File: src/cdp/CdpConnection.ts
Comment (OrKoN) [inline src/cdp/CdpConnection.ts:165]: that would be visible to external clients right? Let's emit this event on an internal emitter only for the mapper implementation

File: src/bidiMapper/modules/cdp/CdpTarget.ts
Comment (sadym-chromium) [reply src/bidiMapper/modules/cdp/CdpTarget.ts:62]: I meant "Enable all required for BiDi Network Module"

Comment (sadym-chromium) [reply src/bidiMapper/modules/cdp/CdpTarget.ts:62]: Actually we don't need this field at all

Comment (sadym-chromium) [reply src/bidiMapper/modules/cdp/CdpTarget.ts:62]: reverted

File: src/bidiMapper/modules/cdp/CdpTarget.ts
Comment (sadym-chromium) [reply src/bidiMapper/modules/cdp/CdpTarget.ts:447]: right, done

File: src/cdp/CdpConnection.ts
Comment (sadym-chromium) [reply src/cdp/CdpConnection.ts:165]: The event will be exposed in `CdpClient` interface and to the clients as `goog:cdp.Page.frameStartedNavigating` event. Are you concerned about the latest, or both?

Comment (OrKoN) [reply src/cdp/CdpConnection.ts:165]: I am concerned that event will be observable by Puppeteer and by clients creating a custom CDP session through Puppeteer who might start depending on it or break due to unknown event/event name format. I do not think even prefixed events should be emitted there. What is the difficulty with using this event internally?

Comment (OrKoN) [reply src/cdp/CdpConnection.ts:165]: Does it even work with custom CDP connections like the one Puppeteer is using?

**Review by OrKoN — COMMENTED** [review]
I see that 'Network.requestWillBeSent' is used once in chromium-bidi? what was the race-condition that you mentioned?

**Comment by OrKoN** [comment]
WDYT if we implement it this way instead https://github.com/GoogleChromeLabs/chromium-bidi/pull/2880 (plus network domain enablement changes )?  Note that TargetEvents.FrameStartedNavigating should always fire before the network event since it is subscribed to the network event first.

**Comment by sadym-chromium** [comment]
> WDYT if we implement it this way instead #2880 (plus network domain enablement changes )? Note that TargetEvents.FrameStartedNavigating should always fire before the network event since it is subscribed to the network event first.

done

**Review by OrKoN — CHANGES_REQUESTED** [review]
EmulatedCdpMapping is not needed anymore?

**Comment by sadym-chromium** [comment]
> EmulatedCdpMapping is not needed anymore?

right, my bad, artefact after merging. Fixed

**Review by OrKoN — APPROVED** [review]

**Comment by OrKoN** [comment]
(No new Puppeteer failures)

**Pull request merged**
